posts/improving-the-access-control-of-a-jhipster-application.md

```diff
@@ -0,0 +1,222 @@
+  </div>
+);
+```
+
+The attribute `hasAnyAuthorities` defines which authorities are required to let the user use the corresponding route.
+
+In order to hide UI elements for a non-admin user, a new property `isAdmin` can be used like below:
+
+```javascript
+render() {
+    const { bankAccountList, match, isAdmin } = this.props;
+    return (
+        <div>
+        [...]
+            {isAdmin && <Button tag={Link} to={`${match.url}/${bankAccount.id}/delete`} color="danger" size="sm">
+                <FontAwesomeIcon icon="trash" />{' '}
+                <span className="d-none d-md-inline">
+                    <Translate contentKey="entity.action.delete">Delete</Translate>
+                </span>
+            </Button>}
+        [...]
+        </div>
+    );
+}
+
+const mapStateToProps = ({ bankAccount, authentication }: IRootState) => ({
+  bankAccountList: bankAccount.entities,
+  isAdmin: hasAnyAuthority(authentication.account.authorities, [AUTHORITIES.ADMIN])
+});
+```
+
+## Angular
+
+In Angular, each module has a file called `modulename.route.ts` that contains the routes definition. Like React, an array of authorities can used in order to control what the user can do.
+
+Here is the route that deletes a `BankAccount` in `bank-account.route.ts`:
+
+```javascript
+export const bankAccountPopupRoute: Routes = [
+    {
+        path: 'bank-account/:id/delete',
+        component: BankAccountDeletePopupComponent,
+        resolve: {
+            bankAccount: BankAccountResolve
+        },
+        data: {
+            authorities: ['ROLE_USER'],
+            pageTitle: 'jhipsterSampleApplicationApp.bankAccount.home.title'
+        },
+        canActivate: [UserRouteAccessService],
+        outlet: 'popup'
+    }
+];
+```
+
+Replacing the authority `ROLE_USER` by `ROLE_ADMIN` will remove the right of deleting a `BankAccount` for a regular user.
+
+In order to display the `delete` button only for admins, the attribute `*jhiHasAnyAuthority="'ROLE_ADMIN'"` can be used like below:
+
+```html
+<button *jhiHasAnyAuthority="'ROLE_ADMIN'"
+        type="submit"
+        [routerLink]="['/', { outlets: { popup: 'bank-account/'+ bankAccount.id + '/delete'} }]"
+        replaceUrl="true"
+        queryParamsHandling="merge"
+        class="btn btn-danger btn-sm">
+    <fa-icon [icon]="'times'"></fa-icon>
+    <span class="d-none d-md-inline" jhiTranslate="entity.action.delete">Delete</span>
+</button>
+```
+
+# Conclusion
+
+Improving the access-control of the API generated by JHipster is pretty easy to achieve, thanks to Spring Security. It can be easily extended and custom authorities can be added to fulfill any specific requirements.
+
+Since the authorities are shared with the UI, the same logic can be done with the frontend's routes to avoid having the regular user accessing unauthorized content. And finally with the help of prebuilt JHipster functions, page elements like link/button can be hidden for regular users and shown for admins.
```

Comment:
```suggestion
Since the authorities are shared with the UI, the same logic can be done with the frontend's routes to avoid having a regular user access unauthorized content. And finally with the help of prebuilt JHipster functions, page elements such as link/button can be hidden for regular users and shown for admins.
```